### DISCUSSION TITLE
Refresh UI on dynamic route param change

### DISCUSSION THREAD

**Discussion by piotrkulpinski** [remix-run/remix#4535] [Q&A] [answered]
Hi,
I'm building an app and having issues refreshing the UI on dynamic route param change. 

I have the route set up like this: `/app/routes/resources/$id.tsx`. 

I want the UI to be fresh every time user navigates to a different resource from the sidebar link. 

https://user-images.githubusercontent.com/358412/200291351-ea0807d3-63f3-4363-abcb-9e23402590a4.mp4

Am I doing something wrong or is there a way to refresh the UI somehow?

Thanks!

**Comment by kiliman** [comment, marked as answer]
Remix doesn't remount your route component on navigation, it's just like fetching new data and re-rendering. However, React will not reset initial state or form input `defaultValue` when re-rendering. You need to force it to unmount.

Typically this is done by setting the `key` prop, perhaps on the form or `<Outlet>`. Use the `useLocation` hook to get the current path and set that as the key.
```ts
const { pathname } = useLocation()

return <Form key={pathname}>...</Form>
```
This will ensure the form is remounted when you change routes. You can also add it to the parent `<Outlet>` which is helpful if you are keeping a lot of local state outside of forms.

**Reply by piotrkulpinski** [reply]
Wow, thanks! Didn't know that setting `key` on any element forces React to remount it on change.

**Reply by kiliman** [reply]
Yes, `key` is essentially the component identity. Change it and React thinks it's a new component, so unmounts the old and mounts a new one.

**Reply by piotrkulpinski** [reply]
@kiliman Is there a way to refresh the state in a route component in a similar manner?

I have a dynamic route with `setState` and I want to make sure I get the fresh state every time the url changes:

```tsx
export default () => {
  const { collection } = useLoaderData<typeof loader>()
  const [state, setState] = useState() // how to refresh this?

  return (
    <Fragment key={collection.id}>
      <Outlet />
    </Fragment>
  )
}
```

**Reply by setoseng** [reply]
> Remix doesn't remount your route component on navigation, it's just like fetching new data and re-rendering. However, React will not reset initial state or form input `defaultValue` when re-rendering. You need to force it to unmount.
> 
> Typically this is done by setting the `key` prop, perhaps on the form or `<Outlet>`. Use the `useLocation` hook to get the current path and set that as the key.
> 
> ```ts
> const { pathname } = useLocation()
> 
> return <Form key={pathname}>...</Form>
> ```
> 
> This will ensure the form is remounted when you change routes. You can also add it to the parent `<Outlet>` which is helpful if you are keeping a lot of local state outside of forms.

This is very helpful. Thank you!

**Comment by znwhite** [comment]
The `key` method is ideal, but as a jankier option (if necessary for your use case) you can always abuse `useEffect` to help you out:

```
const { pathname } = useLocation();
const [showMenu, setShowMenu] = useState(false);

useEffect(() => {
   setShowMenu(false);
}, [pathname])
```
